bftengine/src/bftengine/RequestHandler.cpp

```diff
@@ -117,8 +117,24 @@ void RequestHandler::execute(IRequestsHandler::ExecutionRequestsQueue& requests,
       concord::messages::db_checkpoint_msg::CreateDbCheckpoint createDbChkPtMsg;
       concord::messages::db_checkpoint_msg::deserialize(
           std::vector<std::uint8_t>(req.request, req.request + req.requestSize), createDbChkPtMsg);
-      if (!createDbChkPtMsg.noop)
-        DbCheckpointManager::instance().createDbCheckpointAsync(createDbChkPtMsg.seqNum, timestamp);
+      if (!createDbChkPtMsg.noop) {
+        const auto& lastStableSeqNum = DbCheckpointManager::instance().getLastStableSeqNum();
+        if (lastStableSeqNum == static_cast<SeqNum>(createDbChkPtMsg.seqNum)) {
+          DbCheckpointManager::instance().createDbCheckpointAsync(createDbChkPtMsg.seqNum, timestamp, std::nullopt);
+        } else {
+          // this replica has not reached stable seqNum yet to create snapshot at requested seqNum
+          // add a callback to be called when seqNum is stable. We need to create snapshot on stable
+          // seq num because checkpoint msg certificate is stored on stable seq num and is used for intergrity
+          // check of db snapshots
+          const auto& seqNumToCreateSanpshot = createDbChkPtMsg.seqNum;
+          std::optional snapshotId(DbCheckpointManager::instance().getLastReachableBlock());
```

Comment:
I think this variable can be called `block_id` instead.

Reply:
fixed. Thanks